internal/proxy/ui_requests.go

```diff
@@ -11,212 +11,274 @@ import (
+		nsLogger.Infof("Cached Jenkins route %q in %q", pci.Route, jsessionCookie.Value)
+
+		// If all good, redirect to self to remove token from url
+		nsLogger.Infof("Redirecting to %q", redirectURL)
+		http.Redirect(w, r, redirectURL.String(), http.StatusFound)
+		return
 	}
 
-	if len(r.Cookies()) > 0 { //Check cookies and proxy cache to find user info
+	if len(r.Cookies()) > 0 {
+		// Check cookies and proxy cache to find user info
+		cookieLogger := logger.WithField("part", "cookie")
+
 		for _, cookie := range r.Cookies() {
+
 			cacheVal, ok := p.ProxyCache.Get(cookie.Value)
 			if !ok {
+				cookieLogger.Infof("cookie: %q not in cache", cookie.Name)
+
+				// if the cookie is not in cache but it could be an old idled or jsessionid
+				// cookie so we need to handle that case and clear the cookie
+				if isSessionOrIdledCookie(cookie) {
+					cookieLogger.Infof("cookies: %q is session or idle is OLD; expiring", cookie.Name)
+					expireCookie(w, cookie)
+				}
 				continue
 			}
-			if strings.HasPrefix(cookie.Name, SessionCookie) {
+			cacheKey = cookie.Value
+			pci := cacheVal.(CacheItem)
+			ns = pci.NS
+			clusterURL := pci.ClusterURL
+
+			nsLogger := log.WithFields(log.Fields{"ns": ns, "cluster": clusterURL, "cookie": cookie.Name})
+			nsLogger.Infof("cookie: %q is in cache", cookie.Name)
+
+			if isSessionCookie(cookie) {
 				// We found a session cookie in cache
-				cacheKey = cookie.Value
-				pci := cacheVal.(CacheItem)
-				r.Host = pci.Route //Configure proxy upstream
-				r.URL.Host = pci.Route
-				r.URL.Scheme = pci.Scheme
-				ns = pci.NS
-				needsAuth = false //user is probably logged in, do not redirect
-				noProxy = false
-				break
-			} else if cookie.Name == CookieJenkinsIdled {
-				// Found a cookie saying Jenkins is idled, verify and act accordingly
-				cacheKey = cookie.Value
-				needsAuth = false
-				pci := cacheVal.(CacheItem)
-				ns = pci.NS
-				clusterURL := pci.ClusterURL
+				scLogger := nsLogger.WithField("cookietype", "session")
+				scLogger.Infof("Cache has Jenkins route %q in %q", pci.Route, cookie.Value)
+
+				// ensure jenkins is running
+				state, _, err := p.startJenkins(ns, pci.ClusterURL)
+				if err != nil {
+					p.HandleError(w, err, scLogger)
+					return
+				}
 
+				if state == clients.Running {
+					scLogger.Infof("jenkins is running | so rev proxy: %+v", pci)
+
+					needsAuth = false  // user is logged in; do not redirect
+					okToForward = true // good to be reverse-proxied
+					r.Host = pci.Route // Configure proxy upstream
+					r.URL.Host = pci.Route
+					r.URL.Scheme = pci.Scheme
+					return
+				}
+
+				// we find a session cookie in cache but the pod is not running
+				// so lets clear the cookie and the cache entry
+				p.ProxyCache.Delete(cacheKey)
+				cacheKey = "" // cacheKey isn't valid any more
+				expireCookiesMatching(w, r, isSessionOrIdledCookie)
+				p.recordStatistics(pci.NS, time.Now().Unix(), 0) //FIXME - maybe do this at the beginning?
```

Comment:
Can we create an issue for this and remove the comment?

Reply:
I don't know honestly what the intention here, why are we recording this only here, why not in every return? I will see if there is an issue related to this.